### ISSUE TITLE
replication not making progress?

### ISSUE THREAD

**Issue by bpytlik** [yinqiwen/ardb#213]
I have two ardb instances. I make one the slave of the other, this appears to work but then no progress seems to be being made. Here is the log from the master ARDB:
```
2015-12-09_02:00:34.34155 [593] 12-09 02:00:34,341 INFO [Master]Recv sync command:apsync 374d68139fb84b51651ab52ec509cb1082ba5e90 1 cksm 0
2015-12-09_02:00:34.34156 [593] 12-09 02:00:34,341 DEBUG Detach fd(34) from event loop.
2015-12-09_02:00:34.34157 [593] 12-09 02:00:34,341 DEBUG Process recved cmd[6] cost 119us
2015-12-09_02:00:34.34165 [593] 12-09 02:00:34,341 DEBUG Add slave ip=169.254.0.11,port=6383
2015-12-09_02:00:34.34167 [593] 12-09 02:00:34,341 INFO Partial resynchronization not accepted: Serverkey mismatch (Client asked for '374d68139fb84b51651ab52ec509cb1082ba5e90', I'm '84731d9f23fb932e37db2d992d1741bc703c15dd')
2015-12-09_02:00:34.35120 [593] 12-09 02:00:34,341 INFO [Master]Start dump ardb data to file:/shared/sensors/repl/dump-593-1449626434.ardb
2015-12-09_02:00:35.74686 [593] 12-09 02:00:35,746 INFO [REPL]Saved ardb dump file:/shared/sensors/repl/dump-593-1449626434.ardb
2015-12-09_02:00:35.75396 [593] 12-09 02:00:35,753 INFO [REPL]Send ardb dump file to slave
2015-12-09_02:00:35.84593 [593] 12-09 02:00:35,845 INFO Send complete.
2015-12-09_02:00:35.84595 [593] 12-09 02:00:35,845 INFO [Master]Slave continue sync from backlog offset: 1401
2015-12-09_02:00:35.84596 [593] 12-09 02:00:35,845 DEBUG [Master]Slave sync from 1401 to 1401 at state:10
2015-12-09_02:00:35.84597 [593] 12-09 02:00:35,845 DEBUG [Master]Slave synced.
2015-12-09_02:00:36.84829 [593] 12-09 02:00:36,848 DEBUG [Master]Write command:ping  to slave.
2015-12-09_02:00:36.84832 [593] 12-09 02:00:36,848 DEBUG Ping slaves.
```

And here it is for the slave:
```
2015-12-09_02:00:34.30309 [21171] 12-09 02:00:34,302 DEBUG Process recved cmd[8]:slaveof 169.254.0.10 16379 with flags:0
2015-12-09_02:00:34.30321 [21171] 12-09 02:00:34,302 DEBUG [Slave]Connecting master 169.254.0.10:16379
2015-12-09_02:00:34.30322 [21171] 12-09 02:00:34,302 DEBUG Process recved cmd[8] cost 461us
2015-12-09_02:00:34.30352 [21171] 12-09 02:00:34,303 DEBUG Master conn connected.
2015-12-09_02:00:34.30494 [21171] 12-09 02:00:34,304 INFO [Slave]Remote master is an Ardb instance.
2015-12-09_02:00:34.31007 [21171] 12-09 02:00:34,306 INFO Recv psync reply:FULLRESYNC 84731d9f23fb932e37db2d992d1741bc703c15dd 1401 7725564894718828516
2015-12-09_02:00:48.23152 [21171] 12-09 02:00:48,231 DEBUG Process recved cmd[9]:info replication with flags:0
2015-12-09_02:00:48.23154 [21171] 12-09 02:00:48,231 DEBUG Process recved cmd[9] cost 110us
2015-12-09_02:02:40.32389 [21171] 12-09 02:02:40,323 DEBUG Process recved cmd[10]:info replication with flags:0
2015-12-09_02:02:40.32392 [21171] 12-09 02:02:40,323 DEBUG Process recved cmd[10] cost 109us
2015-12-09_02:03:21.60088 [21171] 12-09 02:03:21,600 DEBUG Process recved cmd[11]:info replication with flags:0
2015-12-09_02:03:21.60090 [21171] 12-09 02:03:21,600 DEBUG Process recved cmd[11] cost 110us
```

Here's the first round of output from `info replication` for the master:
```
# date && redis-cli -p 6383 info replication 
Wed Dec  9 02:05:45 UTC 2015
# Replication
role:master
repl_dir: /shared/sensors/repl
connected_slaves: 1
slave0:ip=169.254.0.11,port=6383,state=online,offset=1401,lag=1449626745, o_buffer_size:0, o_buffer_capacity:0
master_repl_offset: 1835
repl_backlog_size: 52428800
repl_backlog_first_byte_offset: 2
repl_backlog_histlen: 1834
repl_backlog_cksm: 9845643c4bddd646
```

and the slave:
```
# date && redis-cli -p 6383 info replication
Wed Dec  9 02:05:44 UTC 2015
# Replication
role:slave
repl_dir: /shared/sensors/repl
master_host:169.254.0.10
master_port:16379
master_link_status:up
master_last_io_seconds_ago:8
master_sync_in_progress:1
master_sync_left_bytes:4298425677
slave_repl_offset:1
slave_priority:100
connected_slaves: 0
master_repl_offset: 1
repl_backlog_size: 52428800
repl_backlog_first_byte_offset: 2
repl_backlog_histlen: 0
repl_backlog_cksm: 0000000000000000
```

I then waited a while and ran the commands again:
```
Wed Dec  9 02:11:31 UTC 2015
# Replication
role:master
repl_dir: /shared/sensors/repl
connected_slaves: 1
slave0:ip=169.254.0.11,port=6383,state=online,offset=1401,lag=1449627091, o_buffer_size:0, o_buffer_capacity:0
master_repl_offset: 2325
repl_backlog_size: 52428800
repl_backlog_first_byte_offset: 2
repl_backlog_histlen: 2324
repl_backlog_cksm: 353867715d82c029
```

```
Wed Dec  9 02:11:28 UTC 2015
# Replication
role:slave
repl_dir: /shared/sensors/repl
master_host:169.254.0.10
master_port:16379
master_link_status:up
master_last_io_seconds_ago:2
master_sync_in_progress:1
master_sync_left_bytes:4298425677
slave_repl_offset:1
slave_priority:100
connected_slaves: 0
master_repl_offset: 1
repl_backlog_size: 52428800
repl_backlog_first_byte_offset: 2
repl_backlog_histlen: 0
repl_backlog_cksm: 0000000000000000
```

I know doing the replication might take a while, but my concern is that it doesn't look like any progress is being made (at least not that I can tell). I take that mostly from the fact that the master_sync_left_bytes isn't changing.

If there something I'm doing wrong w/ the sync or configuration? Afaict, neither machine is out of resources: cpu, storage, or network bandwidth. Any ideas?

**Comment by bpytlik** [comment]
I should mention this is using ardb version: da4742787867b89560b5e7743a698531bf82d359

**Comment by bpytlik** [comment]
And just to follow up, after about 17 hours, it still appears no progress is being made:
# date && redis-cli -p 6383 info replication
Wed Dec  9 19:48:22 UTC 2015
# Replication
role:slave
repl_dir: /shared/sensors/repl
master_host:169.254.0.10
master_port:16379
master_link_status:up
master_last_io_seconds_ago:3
master_sync_in_progress:1
master_sync_left_bytes:4298425677
slave_repl_offset:1
slave_priority:100
connected_slaves: 0
master_repl_offset: 1
repl_backlog_size: 52428800
repl_backlog_first_byte_offset: 2
repl_backlog_histlen: 0
repl_backlog_cksm: 0000000000000000

**Comment by yinqiwen** [comment]
it seems there is a bug that master send a length '-1', but slave cast it to '4298425677'.

**Comment by bpytlik** [comment]
Is the bug that master is sending a length of -1 or that the slave is miscasting it? I'm asking b/c the issue isn't only in the reporting, afaict, the dbs aren't actually getting synced.

**Comment by yinqiwen** [comment]
Master would send a length header to slave to indicate the length of the synced data later. The length is the file size of dumped file'ardb.dump'.
https://github.com/yinqiwen/ardb/blob/master/src/replication/master.cpp#L444

Slave would keep waiting until the length of  data received from master.
https://github.com/yinqiwen/ardb/blob/master/src/replication/slave.cpp#L329

In this issue, it seems that the dump file is not exist and master use a random value instead of 0, which is the bug IMO.

**Comment by yinqiwen** [comment]
fixed in 0.9
